File: rpg-encoder/tests/fixtures/lua/ffi.lua
local ffi = require("ffi")

ffi.cdef[[
    int add_numbers(int a, int b);
    int process_data(const void* data, size_t len);
    void* malloc(size_t size);
    void free(void* ptr);
    size_t strlen(const char* s);
]]

local mylib = ffi.load("mylib")

local function addViaFfi(a, b)
    return mylib.add_numbers(a, b)
end

local function processViaFfi(data)
    local cdata = ffi.new("char[?]", #data + 1)
    ffi.copy(cdata, data)
    return mylib.process_data(cdata, #data)
end

local function useStdLib(s)
    return ffi.C.strlen(s)
end

local function allocateBuffer(size)
    local ptr = ffi.C.malloc(size)
    return ptr
end

local int_ptr = ffi.new("int[10]")
local struct_ptr = ffi.new("struct { int x; int y; }")

return {
    addViaFfi = addViaFfi,
    processViaFfi = processViaFfi,
    useStdLib = useStdLib
}

File: rpg-encoder/tests/fixtures/ruby/ffi.rb
require 'ffi'

module NativeLib
  extend FFI::Library
  ffi_lib 'libmylib.so'

  attach_function :add_numbers, [:int, :int], :int
  attach_function :process_data, [:pointer, :size_t], :int
  
  callback :data_callback, [:pointer, :int], :void
  attach_function :register_callback, [:data_callback], :void
end

module NativeUtils
  extend FFI::Library
  ffi_lib 'libc.so.6'
  
  attach_function :malloc, [:size_t], :pointer
  attach_function :free, [:pointer], :void
end

def add_via_ffi(a, b)
  NativeLib.add_numbers(a, b)
end
